### PR TITLE
fix(1865): [2] Remove build step

### PR DESCRIPTION
**Pull request by sugarnaoming**
## Context
As mentioned here screwdriver-cd/screwdriver#1865 , `builds.steps` and `steps.command` are both same data.
And these steps are defined by user's yaml and it tends to have a big size.

## Objective
- Use `stepFactory` to get steps instead of `buildFactory`.

## References
#1865 

## License
I confirm that this contribution is made under a BSD license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by kkisic** [comment]
Need to fix lint

File: plugins/builds/get.js [hunk 1]
```diff
@@ -23,27 +23,32 @@ module.exports = () => ({
...
 
-            return factory.get(request.params.id)
-                .then((model) => {
-                    if (!model) {
+            return Promise.all([buildFactory.get(request.params.id),
+                stepFactory.list({ params: { buildId: request.params.id } })])
+                .then(([buildModel, stepsModel]) => {
+                    if (!buildModel) {
                         throw boom.notFound('Build does not exist');
                     }
```
Comment (kkisic) [inline plugins/builds/get.js]: Do we need to check `stepModel` as same as `buildModel` ?

File: test/plugins/builds.test.js [hunk 2]
```diff
@@ -2728,41 +2726,14 @@ describe.only('build plugin test', () => {
...
-                assert.deepProperty(reply.result, 'name', 'test');
-                assert.deepProperty(reply.result, 'code', 0);
-                assert.deepProperty(reply.result, 'endTime', options.payload.endTime);
-                assert.notDeepProperty(reply.result, 'startTime');
-            });
-        });
-
-        it('returns 200 when updating the code/endTime when the step model exists', () => {
+        it('returns 200 when updating the code/endTime when the step model exists', () =>
             server.inject(options).then((reply) => {
```
Comment (kkisic) [inline test/plugins/builds.test.js]: Do we need to add return statement?

File: plugins/builds/get.js [hunk 3]
```diff
@@ -23,27 +23,40 @@ module.exports = () => ({
...
 
-                    if (Array.isArray(model.environment)) {
-                        return reply(model.toJson());
+
+                    // This if statement should be removed after enough time has passed since build.steps removed.
+                    // Make orders of steps in completed builds sure,
+                    // because steps in old builds in DB have the order not sorted.
+                    if (buildModel.endTime) {
+                        stepsModel.sort((a, b) => new Date(a.startTime).getTime() - new Date(b.startTime).getTime());
+                    }
```
Comment (kkisic) [inline plugins/builds/get.js]: Do we need to add the test case about these codes?

File: test/plugins/builds.test.js (hunk 2 above)
Comment (sugarnaoming) [reply test/plugins/builds.test.js]: This function is written in arrow function without braces.

Comment (kkisic) [reply test/plugins/builds.test.js]: I see.

File: plugins/builds/get.js (hunk 1 above)
Comment (sugarnaoming) [reply plugins/builds/get.js]: Thank you.
I fixed

File: plugins/builds/get.js (hunk 3 above)
Comment (sugarnaoming) [reply plugins/builds/get.js]: Thank you.
added test

File: plugins/builds/get.js (hunk 1 above)
Comment (kkisic) [reply plugins/builds/get.js]: Do we need to add test cases for empty `stepsModel`?

File: plugins/builds/get.js (hunk 3 above)
Comment (kkisic) [reply plugins/builds/get.js]: Would you mind adding some test code for the case where `stepsModel` are in the order of `id` when `buildModel.endTime` is undefined?

File: plugins/builds/steps/logs.js [hunk 4]
```diff
@@ -160,21 +160,13 @@ module.exports = config => ({
             }
         },
         handler: (req, reply) => {
-            const factory = req.server.app.buildFactory;
+            const factory = req.server.app.stepFactory;
```
Comment (tk3fftk) [inline plugins/builds/steps/logs.js]: This should be renamed to `stepFactory` like `plugins/builds/get.js` for refactor.

File: plugins/builds/get.js [hunk 5]
```diff
@@ -23,27 +23,28 @@ module.exports = () => ({
...
+            return buildFactory.get(request.params.id)
+                .then((buildModel) => {
+                    if (!buildModel) {
                         throw boom.notFound('Build does not exist');
                     }
 
-                    if (Array.isArray(model.environment)) {
-                        return reply(model.toJson());
+                    if (Array.isArray(buildModel.environment)) {
+                        return reply(buildModel.toJsonWithSteps());
```
Comment (tk3fftk) [inline plugins/builds/get.js]: How about another `build.toJson()` cases like this line? Do we need adding steps same with here after removing `builds.steps`?
https://github.com/screwdriver-cd/screwdriver/blob/b9045682bcc43af1846a9cd9d15400fd79c5bdc5/plugins/builds/update.js#L184

File: plugins/builds/steps/logs.js (hunk 4 above)
Comment (yuichi10) [reply plugins/builds/steps/logs.js]: I fixed

File: plugins/builds/get.js (hunk 5 above)
Comment (yuichi10) [reply plugins/builds/get.js]: I fixed.

**Comment by s-yoshika** [comment]
New steps endpoint seems appeared. We should fix it.
https://github.com/sonic-screwdriver-cd/screwdriver/tree/master/plugins/builds/steps

File: plugins/builds/steps/get.js [hunk 6]
```diff
@@ -21,18 +21,10 @@ module.exports = () => ({
...
-                    if (!buildModel) {
-                        throw boom.notFound('Build does not exist');
-                    }
-
-                    const stepModel = buildModel.steps.filter(step => (
-                        step.name === request.params.name
-                    )).pop();
+            const stepFactory = request.server.app.stepFactory;
 
+            stepFactory.get({ buildId: request.params.id, name: request.params.name })
```
Comment (tkyi) [inline plugins/builds/steps/get.js]:
```suggestion
            return stepFactory.get({ buildId: request.params.id, name: request.params.name })
```

should we add `return`?

File: plugins/builds/steps/logs.js [hunk 7]
```diff
@@ -160,21 +160,13 @@ module.exports = config => ({
...
-                .then((model) => {
-                    if (!model) {
-                        throw boom.notFound('Build does not exist');
-                    }
-
-                    const stepModel = model.steps.filter(step => (
-                        step.name === stepName
-                    )).pop();
-
+            stepFactory.get({ buildId, name: stepName })
```
Comment (tkyi) [inline plugins/builds/steps/logs.js]:
```suggestion
            return stepFactory.get({ buildId, name: stepName })
```
same here?

**Comment by yuichi10** [comment]
@tkyi  Thank you for your suggestion.
